spec/20-http_request_header_format.md

```diff
@@ -80,22 +81,25 @@ The dash (`-`) character is used as a delimiter between fields.
 #### version
 
 ``` abnf
-version         = 2HEXDIGLC   ; this document assumes version 00. Version ff is forbidden
+version         = 2HEXDIGLC   ; this document assumes version 01. Version ff is forbidden
 ```
 
 The value is US-ASCII encoded (which is UTF-8 compliant).
 
-Version (`version`) is 1 byte representing an 8-bit unsigned integer. Version `ff` is invalid. The current specification assumes the `version` is set to `00`.
+Version (`version`) is 1 byte representing an 8-bit unsigned integer. Version `ff` is invalid. The current specification assumes the `version` is set to `01`.
 
 #### version-format
 
-The following `version-format` definition is used for version `00`.
+The following `version-format` definition is used for version `01`.
 
 ``` abnf
-version-format   = trace-id "-" parent-id "-" trace-flags
-trace-id         = 32HEXDIGLC  ; 16 bytes array identifier. All zeroes forbidden
-parent-id        = 16HEXDIGLC  ; 8 bytes array identifier. All zeroes forbidden
-trace-flags      = 2HEXDIGLC   ; 8 bit flags. Currently, only one bit is used. See below for details
+version-format         = trace-id "-" parent-id "-" trace-flags [ "-" sampling-constant ]
+trace-id               = 32HEXDIGLC  ; 16 bytes array identifier. All zeroes forbidden
+parent-id              = 16HEXDIGLC  ; 8 bytes array identifier. All zeroes forbidden
+trace-flags            = 2HEXDIGLC   ; 8 bit flags. Currently, only one bit is used. See below for details
+sampling-constant      = sampling-random-value parent-sampling-constant
+sampling-random-value  = 2HEXDIGLC   ; geometrically distributed 8-bit random value used for consistent sampling
+parent-sampling-constant = 2HEXDIGLC   ; 8-bit value used to represent the head sampling probability.
```

Comment:
I think the intention of limiting values to 6-bits only was that on the wire the combined constant could be represented with just 3 hexes, not 4. Was this discussed and decided against?

Reply:
I understood the original 6-bit requirement was because of the desire to use a 64-bit int, but due to some restrictions of some languages (e.g. java) it is easier to ignore the sign bit hence 63 bits of data. Then a state for 0 was needed because there is no `x` such that `2^-x = 0` leaving 62 bits. From there, only 6 bits is needed to represent the number of leading 0s in a 62-bit binary number and 62 was the highest required exponent because `2^-62` was the minimum sampling rate.

AFAIK the 6-bit + 6-bit 3-hex-character representation was just a happy coincidence of this earlier decision.

In any case, the decision to use 4 or 3 characters doesn't matter that much to me. 4 is slightly easier to parse (parse 2 hex bytes vs some bit shifting needed) and much easier for a human to read.

Reply:
I agree on both counts.

The use of 4 bytes is simpler to explain and implement, and 3 is definitely acceptable.
One reason to go with 4 instead of 3 is that it makes #467 orthogonal. The random value is only needed when the randomness flag is not set, according to that proposal (@bogdandrutu). Combining this proposal with #467, the `p` value is only needed _when sampled_ and the `r` value is built in--we arrive at a nice outcome:

- With randomness flag set: 2 bytes extra when sampled, 0 bytes extra when not sampled 
- Without randomness flag set: 4 bytes extra when sampled, 2 bytes extra when not sampled
-